text/metrics/0185-uptime-monitoring-semantic-conventions.md

```diff
@@ -0,0 +1,104 @@
+
+This use case can also be met with a restart count metric, but "up and to the right" uptime graphs are more comfortable for users.
+
+#### Alerting on restarts using uptime
+
+Some customers want to configure an alert whenever their process (re)starts. Typically, they do this by configuring an alert that fires when the absolute value of the uptime (time since process start) is below a certain threshold.
+
+This type of alert is considered bad practice because it is, by definition, non-actionable. By the time the uptime is reported as 0, the process has already started and is now running. The alert will self-close as soon as the process has been running for a few minutes.
+
+We recommend in general not configuring alerts on this type of metric, but it is a common practice.
+
+#### Alerting on current process health
+
+Some customers want to configure an alert when a process is unhealthy, or the number of healthy processes falls below a certain level. They can do this using a boolean metric exported from each task that reports if it is currently healthy.
+
+In general, we recommend that services prefer exposing more detailed metrics instead of boolean health. For example, a user should prefer an alert defined as "successful qps > 0.1" instead of an alert defined as "healthy = TRUE". The reason for this is that boolean health metrics essentially move the alert condition definition inside the process, removing the user's ability to control those conditions.
+
+#### Alerting on crashing processes
+
+Users want to know if their processes are restarting frequently (for example, as part of a crash loop). This is the hardest type of metric to report, because restarting processes may not be healthy enough to report that they are crashing.
+
+When there is some kind of supervisor process that is separate from the process being monitored, this can be tracked by having the supervisor process report the number of times a process has restarted. Then an alert can be configured when the rate of that count exceeds a threshold (e.g. "process has restarted ≥ 5 times in one hour").
+
+Since supervisor processes like this do not often exist, other ways to partially achieve this are to configure an alert on the average uptime over a long time period (e.g. "average uptime over 60m window < 15m"), and/or alert when a metric is absent.
+
+## Internal details
+
+We propose the following metrics be used to track uptime within OpenTelemetry:
+
+| Name                   | Description                  | Units | Instrument Type              |
+| ---------------------- | ---------------------------- | ----- | -----------------------------|
+| *.uptime               | Seconds since last restart   | s     | Asynchronous Gauge           |
+| *.health               | Availability flag.           | 1     | Asynchronous Gauge           |
+| *.restart_count        | Number of restarts.          | 1     | Asynchronous Counter         |
```

Comment:
There is already a `restart_count` _attribute_ defined for resources: https://github.com/open-telemetry/opentelemetry-specification/blob/a25d5f03ab58ecf88c09f635df97d2328b5ba237/specification/resource/semantic_conventions/k8s.md#container
It would be useful to tell how these two are related (if they are).

Reply:
From a metric standpoint, having restart_count as a resource attribute is really really bad.  I'm suprised no one commented on that, but I think it should be dropped, as it violates resource identity and causes high cardinality.


The TL;DR: though is that restart_count is something you'd want to alert on, and for metric-systems that means it needs to be a metric.  They'd likely be the *same* value, but one is a resource attribute the other is a metric data point.

Reply:
> From a metric standpoint, having restart_count as a resource attribute is really really bad. I'm suprised no one commented on that, but I think it should be dropped, as it violates resource identity and causes high cardinality.

It was discussed, see https://github.com/open-telemetry/opentelemetry-specification/pull/1945#pullrequestreview-759394466

The conclusion was that for a k8s container in a pod it is an identifying attribute, not a metric.

Reply:
The Kubernetes restart count is not actually part of a unique-across-time identifier:

https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.21/#containerstatus-v1-core

> The number of times the container has been restarted, currently based on the number of dead containers that have not yet been removed. Note that this is calculated from dead containers. But those containers are subject to garbage collection. This value will get capped at 5 by GC.

Arguably that means it's not even useful as a metric, but it's certainly not a unique identifier.